src/sequence_state.cc

```diff
@@ -226,55 +288,62 @@ SequenceStates::OutputState(
   std::advance(input_states_itr, iter_advance);
   auto& input_state_r = input_states_[input_states_itr->first];
+  bool reuse_buffer = output_states_[name]->ReuseBuffer();
 
   if (output_state != nullptr) {
     *output_state = output_states_[name].get();
   }
 
-  output_state_r->SetStateUpdateCallback([&output_state_r, &input_state_r]() {
-    // Swap the internal memory if the size of the input and output state is
-    // equal
-
-    if (output_state_r->Data()->TotalByteSize() ==
-        input_state_r->Data()->TotalByteSize()) {
-      std::shared_ptr<Memory> temp_memory = input_state_r->Data();
-      RETURN_IF_ERROR(input_state_r->RemoveAllData());
-      RETURN_IF_ERROR(input_state_r->SetData(output_state_r->Data()));
-      RETURN_IF_ERROR(output_state_r->RemoveAllData());
-      RETURN_IF_ERROR(output_state_r->SetData(temp_memory));
-    } else {
-      // If the size of output state is different from the input state, allocate
-      // a new memory for the input state with the same size as output state.
-      TRITONSERVER_MemoryType memory_type;
-      int64_t memory_type_id;
-
-      const std::shared_ptr<MutableMemory>& input_memory =
-          reinterpret_cast<const std::shared_ptr<MutableMemory>&>(
-              input_state_r->Data());
-
-      input_memory->MutableBuffer(&memory_type, &memory_type_id);
-      std::shared_ptr<AllocatedMemory> memory =
-          std::make_shared<AllocatedMemory>(
-              output_state_r->Data()->TotalByteSize(), memory_type,
-              memory_type_id);
-      RETURN_IF_ERROR(input_state_r->RemoveAllData());
-      RETURN_IF_ERROR(input_state_r->SetData(output_state_r->Data()));
-      RETURN_IF_ERROR(output_state_r->RemoveAllData());
-      RETURN_IF_ERROR(output_state_r->SetData(memory));
-    }
-
-    // Update the shape and data type of the output state if it doesn't match
-    // the input state.
-    if (input_state_r->Shape() != output_state_r->Shape()) {
-      *input_state_r->MutableShape() = output_state_r->Shape();
-    }
-
-    if (input_state_r->DType() != output_state_r->DType()) {
-      *input_state_r->MutableDType() = output_state_r->DType();
-    }
-
-    return Status::Success;
-  });
+  output_state_r->SetStateUpdateCallback(
+      [&output_state_r, &input_state_r, reuse_buffer]() {
+        // Swap the internal memory if the size of the input and output state is
+        // equal
+
+        if (output_state_r->Data()->TotalByteSize() ==
+            input_state_r->Data()->TotalByteSize()) {
+          if (!reuse_buffer) {
+            std::shared_ptr<Memory> temp_memory = input_state_r->Data();
+            RETURN_IF_ERROR(input_state_r->RemoveAllData());
+            RETURN_IF_ERROR(input_state_r->SetData(output_state_r->Data()));
+            RETURN_IF_ERROR(output_state_r->RemoveAllData());
+            RETURN_IF_ERROR(output_state_r->SetData(temp_memory));
+          }
+        } else {
+          if (!reuse_buffer) {
```

Comment:
see above - seems like we don't execute either the if or else clause if we are reusing the buffer - can we move that check outside?